mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mjpc/planners/cem/planner.h"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <mutex>
+#include <shared_mutex>
+
+#include <absl/random/random.h>
+#include <mujoco/mujoco.h>
+#include "mjpc/array_safety.h"
+#include "mjpc/planners/cem/policy.h"
+#include "mjpc/states/state.h"
+#include "mjpc/trajectory.h"
+#include "mjpc/utilities.h"
+
+namespace mjpc {
+
+namespace mju = ::mujoco::util_mjpc;
+
+// initialize data and settings
+void CEMPlanner::Initialize(mjModel* model, const Task& task) {
+  // delete mjData instances since model might have changed.
+  data_.clear();
+  // allocate one mjData for nominal.
+  ResizeMjData(model, 1);
+
+  // model
+  this->model = model;
+
+  // task
+  this->task = &task;
+
+  // rollout parameters
+  timestep_power = 1.0;
+
+  // sampling noise
+  noise_exploration = GetNumberOrDefault(0.1, model, "sampling_exploration");  // only controls initial variance
+
+  // set number of trajectories to rollout
+  num_trajectory_ = GetNumberOrDefault(10, model, "sampling_trajectories");
+
+  // set number of elite samples
+  n_elites = GetNumberOrDefault(num_trajectory_ / 2, model, "n_elites");  // half rounded down
+
+  if (num_trajectory_ > kMaxTrajectory) {
+    mju_error_i("Too many trajectories, %d is the maximum allowed.",
+                kMaxTrajectory);
+  }
+
+  winner = 0;
+}
+
+// allocate memory
+void CEMPlanner::Allocate() {
+  // initial state
+  int num_state = model->nq + model->nv + model->na;
+
+  // state
+  state.resize(num_state);
+  mocap.resize(7 * model->nmocap);
+  userdata.resize(model->nuserdata);
+
+  // policy
+  int num_max_parameter = model->nu * kMaxTrajectoryHorizon;
+  policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+  previous_policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+
+  // scratch
+  parameters_scratch.resize(num_max_parameter);
+  times_scratch.resize(kMaxTrajectoryHorizon);
+
+  // noise
+  noise.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  stdev.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  fill(stdev.begin(), stdev.end(), noise_exploration);  // TODO: if this gets called before Initialize, then move this to above
+
+  // trajectory and parameters
+  winner = -1;
+  for (int i = 0; i < kMaxTrajectory; i++) {
+    trajectory[i].Initialize(num_state, model->nu, task->num_residual,
+                             task->num_trace, kMaxTrajectoryHorizon);
+    trajectory[i].Allocate(kMaxTrajectoryHorizon);
+    candidate_policy[i].Allocate(model, *task, kMaxTrajectoryHorizon);
+  }
+
+  // need to initialize an arbitrary order of the trajectories
+  trajectory_order.reserve(num_trajectory_);
+  for (int i = 0; i < num_trajectory_; i++) {
+    trajectory_order.push_back(i);
+  }
```

Comment:
new - if we don't do this, then the first iteration, there is no trajectory_order so indexing into it to run the first CEM step segfaults.